### ISSUE TITLE
how to use shap for custom-built models

### ISSUE THREAD

**Issue by cliffrunner** [shap/shap#457]
Hello,

I have an ensemble model that returns weighted average of results from elasticnet/ridge/xgboost/randomforest. I understand I can use linear explainer for elasticnet and ridge, and tree explainer for xgbost and randomforest, but is there a way I can combine these to get a shap score?

thank you.

**Comment by AllardJM** [comment]
Wont KernelExplainer work in this case - you just need to script a predict function. The model itself is not important.

**Comment by cliffrunner** [comment]
True. But KernelExplainer takes forever to finish for my case compared to linear and tree explainers

**Comment by AllardJM** [comment]
Perhaps a weighted average of the Shap values would be suitable (weighted in the same way as the ensemble output), When I have faced this situation, I decided to use the ensemble for prediction but to select the best single model for interpretation and explanation. For my use case that was sufficient.

**Comment by slundberg** [comment]
It turns out that a linear combination of models over the same set of input features has SHAP values that are the same linear combination of the SHAP values of each individual model (linearity is one of the properties of Shapley values). So yes, if you are really just linearly combining the output of the model (in the log odds space if you are doing classification), then just use the same weighted average of each of the individual explanations.

**Comment by melzzzzzzen** [comment]
> It turns out that a linear combination of models over the same set of input features has SHAP values that are the same linear combination of the SHAP values of each individual model (linearity is one of the properties of Shapley values). So yes, if you are really just linearly combining the output of the model (in the log odds space if you are doing classification), then just use the same weighted average of each of the individual explanations.

What if the model output of two individual predictions are multiplied together instead of summed? 
Based on SHAP values, mSHAP values have recently been proposed for this purpose, however this is an extension of Tree SHAP, and I am interested in using Deep SHAP. Is it possible using the shap package to find the SHAP values of two Neural Network predictions multiplied?

**Comment by SimonVedder** [comment]
> 

How does this work if we are using a linear combination of two models that use different features?

**Comment by Haonan917** [comment]
> How does this work if we are using a linear combination of two models that use different features?

Do you have answers to them? These are also my questions.
